Replicator/Inserter.cc

```diff
@@ -191,25 +191,47 @@ namespace litecore { namespace repl {
     // Callback from c4doc_put() that applies a delta, during _insertRevisionsNow()
     C4SliceResult Inserter::applyDeltaCallback(C4Document *c4doc,
                                                C4Slice deltaJSON,
+                                               C4RevisionFlags *revFlags,
                                                C4Error *outError)
     {
         Doc doc = _db->applyDelta(c4doc, deltaJSON, true);
         alloc_slice body = doc.allocedData();
+        Dict root = doc.root().asDict();
+        FLSharedKeys sk = nullptr;
+        bool bodyChanged = false;
 
         if (!_db->disableBlobSupport()) {
             // After applying the delta, remove legacy attachment properties and any other
             // "_"-prefixed top level properties:
-            Dict root = doc.root().asDict();
             if (C4Document::hasOldMetaProperties(root)) {
                 body = nullslice;
                 try {
-                    FLSharedKeys sk = _db->insertionDB().useLocked()->getFleeceSharedKeys();
+                    sk = _db->insertionDB().useLocked()->getFleeceSharedKeys();
                     body = C4Document::encodeStrippingOldMetaProperties(root, sk);
+                    bodyChanged = true;
                 } catchAndWarn();
                 if (!body)
                     *outError = C4Error::make(WebSocketDomain, 500, "invalid legacy attachments");
             }
         }
+        if ( body && revFlags != nullptr ) {
+            bool hasAttachments = false;
+            if ( bodyChanged ) {
+                doc = Doc(body, kFLTrusted, sk);
+                root = doc.asDict();
+            }
+            // This is not optimal. Consider to add a method to DBAccess, hasBlobReferences
+            _db->findBlobReferences(root, true, [&](FLDeepIterator, Dict, const C4BlobKey&) {
+                hasAttachments = true;
+            });
+            if ( hasAttachments && !(*revFlags & kRevHasAttachments) ) {
+                *revFlags |= kRevHasAttachments;
+            } else if ( !hasAttachments && (*revFlags & kRevHasAttachments) ) {
+                // This shouldn't happen
```

Comment:
Why shouldn't it happen?  What if all the attachments were removed by the incoming rev?

Reply:
As per my current reading, there are only two places where the flag is set. 1. when _mayContainBlobs == true. In this case. we apply the delta first. Then, we read in the doc to find all blobs. That is, we look at the final whole document body of the rev to be put nto the store. 2. when _mayContainBlobs == false, in this case we also read the doc after the delta applied. Thees two paths do not intersect. Unless the new rev to be inserted is initialized with the flag being set, we cannot have this case. My current understanding is, the attachments are not inherited from the parent revision. Like other properties, they have to be in the body of the new revision. To remove it, we simply not include that attachment in the new revision body.
I could be wrong, so it is a DebugAssert. Won't affect the final released program. Only I think it would never happen.

Reply:
So you mean to say that these flags are not initialized to what the parent revision flags were?  If so I can see the reasoning.